File: 프로그래머스/brute_force/전력망을 둘로 나누기.py
from collections import defaultdict

def solution(n, wires):
    answer = n
    tree = defaultdict(list)
    
    def dfs(k, visited):
      visited.append(k)
      
      for t in tree[k]:
        if t in visited: 
          continue
      
        dfs(t, visited)
    
    
    for w in wires:
        v1, v2 = w
        tree[v1].append(v2)
        tree[v2].append(v1)
    
    for k in tree.keys():
      for v in tree[k]:
        visited1 = [k]
        visited2 = [v]
        dfs(v, visited1)
        dfs(k, visited2)
        
        gap = abs((len(visited1) - 1) - (len(visited2)-1))
        answer = min(answer, gap)
              
    return answer
  
print(solution(7, [[1,2],[2,7],[3,7],[3,4],[4,5],[6,7]]))

File: 프로그래머스/brute_force/전력망을 둘로 나누기.js
function solution(n, wires) {
  var answer = n;

  const graph = Array.from({ length: n }, (_, i) => i + 1).reduce((acc, x) => {
    acc[x] = [];
    return acc;
  }, {});

  wires.forEach(([a, b]) => {
    graph[a].push(b);
    graph[b].push(a);
  });

  function dfs(node, visited) {
    visited.push(node);

    for (let n of graph[node]) {
      if (!visited.includes(n)) {
        dfs(n, visited);
      }
    }

    return visited.length;
  }

  for (let w of wires) {
    const [a, b] = w;
    const left = dfs(a, [a, b]);
    const right = dfs(b, [a, b]);
    answer = Math.min(answer, Math.abs(left - right));
  }

  return answer;
}
